File: src/core/FontStore.java
package core;

import java.util.HashMap;
import java.util.Map;

import org.newdawn.slick.AngelCodeFont;
import org.newdawn.slick.Font;
import org.newdawn.slick.SlickException;

/**
 * Manages fonts for use in the game.
 */
public class FontStore {
	public enum FontID {
		H1,
		H2,
		FIELD,
		BIG_FIELD
	}
	
	private static FontStore fontStore;
	
	public final Map<String, Font> FONTS;
	
	/**
	 * Constructs a new {@code FontStore} and initializes it.
	 */
	private FontStore() {
		fontStore = this;
		
		FONTS = new HashMap<String, Font>();
		
		try {
			initialize();
		} catch (SlickException e) {
			e.printStackTrace();
		}
	}
	
	/**
	 * Initializes the store with all the images.
	 * @throws SlickException
	 */
	private void initialize() throws SlickException {
		FONTS.put(FontID.H1.toString(), new AngelCodeFont("resources/fonts/04b03_h1.fnt", "resources/fonts/04b03_h1.png", false));
		FONTS.put(FontID.H2.toString(), new AngelCodeFont("resources/fonts/04b03_h2.fnt", "resources/fonts/04b03_h2.png", false));
		FONTS.put(FontID.FIELD.toString(), new AngelCodeFont("resources/fonts/04b03_field.fnt", "resources/fonts/04b03_field.png", false));
	}
	
	/**
	 * Get the {@code Font} for the key.
	 * @param name Key for the font
	 * @return Font at that key, if any
	 */
	public Font getFont(FontID font) {
		return FONTS.get(font.toString());
	}
	
	/**
	 * Returns the {@code FontStore} instance. Creates one if one does not exist.
	 * @return Instance of FontStore
	 */
	public static FontStore get() {
		if (fontStore == null) {
			fontStore = new FontStore();
		}
		
		return fontStore;
	}
}

File: web/src/core/FontStore.ts
import { BitmapFont } from './BitmapFont';

export enum FontID {
  H1 = "H1",
  H2 = "H2",
  FIELD = "FIELD",
}

class FontStoreClass {
  private fonts: Map<FontID, BitmapFont> = new Map();
  private initialized: boolean = false;

  async initialize(): Promise<void> {
    if (this.initialized) return;

    const fontConfigs = [
      { id: FontID.H1, fnt: '/fonts/04b03_h1.fnt', png: '/fonts/04b03_h1.png' },
      { id: FontID.H2, fnt: '/fonts/04b03_h2.fnt', png: '/fonts/04b03_h2.png' },
      { id: FontID.FIELD, fnt: '/fonts/04b03_field.fnt', png: '/fonts/04b03_field.png' },
    ];

    await Promise.all(
      fontConfigs.map(async (config) => {
        const font = new BitmapFont();
        await font.load(config.fnt, config.png);
        this.fonts.set(config.id, font);
      })
    );

    this.initialized = true;
  }

  getFont(id: FontID): BitmapFont {
    const font = this.fonts.get(id);
    if (!font) {
      throw new Error(`Font ${id} not loaded`);
    }
    return font;
  }
}

export const FontStore = new FontStoreClass();
